<a href="https://colab.research.google.com/github/julietcass71/APRENDIZAJE-AUTOMATICO/blob/main/regresion_multiple_presion_arterial_Public.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Ejercicio de regresión lineal - Nancy Julieta Cassano**

**Objetivo:**
Construir un modelo de regresión lineal múltiple para predecir la presión arterial de las personas en función de diferentes variables explicativas.

**Dataset:** dataset_regresion_multiple.csv

**Columnas incluidas:**

**-edad:** Edad de la persona (años).

**-horas_ejercicio:** Horas semanales de ejercicio físico.

**-peso:** Peso corporal (kg).

**-estres:** Nivel de estrés percibido (1 a 10).

**-ingresos:** Ingreso anual en pesos.

**-horas_tv:** Horas semanales viendo televisión.

**-presion_arterial:** Variable objetivo (mmHg).

**Consignas**

**1. Exploración inicial:**

-Describir el dataset (medias, máximos, mínimos, correlaciones).

-Identificar qué variables podrían estar más relacionadas con la presión arterial.

**2. Modelo de regresión lineal múltiple:**

Estimar un modelo con todas las variables predictoras.

**3. Selección de variables:**

-Determinar qué variables tienen mayor peso explicativo en la presión arterial.

-Comparar el modelo con todas las variables vs. uno simplificado con las más relevantes (analizar correlación por ejemplo u otro parámetro que consideres)

**4. Evaluación del modelo:**

-Calcular el R², RMSE y MAE.

-Interpretar si el modelo ajusta adecuadamente.

**5. Informe final:**

-Concluir cuáles factores influyen realmente en la presión arterial.

-Discutir la importancia de edad, peso, estrés y ejercicio frente a ingresos y horas de TV.

In [1]:
!pip -q install cookiecutter
!cookiecutter -c v1 https://github.com/drivendata/cookiecutter-data-science


  [1/8] project_name (project_name): Regresión múltiple presión arterial
  [2/8] repo_name (regresión_múltiple_presión_arterial): regresion-multiple-presion-arterial
  [3/8] author_name (Your name (or your organization/company/team)): Nancy Julieta Cassano
  [4/8] description (A short description of the project.): Modelo de regresión lineal múltiple para predecir la presión arterial a partir de edad, peso, ejercicio, estrés, ingresos y horas de TV.
  [5/8] Select open_source_license
    1 - MIT
    2 - BSD-3-Clause
    3 - No license file
    Choose from [1/2/3] (1): 3
  [6/8] s3_bucket ([OPTIONAL] your-bucket-for-syncing-data (do not include 
's3://')): 
  [7/8] aws_profile (default): 
  [8/8] Select python_interpreter
    1 - python3
    2 - python
    Choose from [1/2] (1): 1


*** DEPRECATION WARNING ***

Cookiecutter data science is moving to v2 soon, which will entail using
the command `ccds ...` rather than `cookiecutter ...`. The cookiecutter command
will continue to work, and 

In [2]:
%cd /content/regresion-multiple-presion-arterial
!pwd
!ls -la


/content/regresion-multiple-presion-arterial
/content/regresion-multiple-presion-arterial
total 76
drwxr-xr-x 9 root root 4096 Sep 25 00:35 .
drwxr-xr-x 1 root root 4096 Sep 25 00:35 ..
drwxr-xr-x 6 root root 4096 Sep 25 00:35 data
drwxr-xr-x 2 root root 4096 Sep 25 00:35 docs
-rw-r--r-- 1 root root  459 Sep 25 00:35 .env
-rw-r--r-- 1 root root 1003 Sep 25 00:35 .gitignore
-rw-r--r-- 1 root root    1 Sep 25 00:35 LICENSE
-rw-r--r-- 1 root root 4703 Sep 25 00:35 Makefile
drwxr-xr-x 2 root root 4096 Sep 25 00:35 models
drwxr-xr-x 2 root root 4096 Sep 25 00:35 notebooks
-rw-r--r-- 1 root root 3041 Sep 25 00:35 README.md
drwxr-xr-x 2 root root 4096 Sep 25 00:35 references
drwxr-xr-x 3 root root 4096 Sep 25 00:35 reports
-rw-r--r-- 1 root root  103 Sep 25 00:35 requirements.txt
-rw-r--r-- 1 root root  333 Sep 25 00:35 setup.py
drwxr-xr-x 6 root root 4096 Sep 25 00:35 src
-rw-r--r-- 1 root root  632 Sep 25 00:35 test_environment.py
-rw-r--r-- 1 root root   50 Sep 25 00:35 tox.ini


In [3]:
from google.colab import files
files.upload()


Saving dataset_regresion_multiple_TA.csv to dataset_regresion_multiple_TA.csv


{'dataset_regresion_multiple_TA.csv': b'edad;horas_ejercicio;peso;estres;ingresos;horas_tv;presion_arterial\n58;7;81;1;21150;12;73.83\n48;2;55;1;94740;22;82.26\n34;2;81;3;86617;24;90.47\n62;0;53;6;36896;34;93.49\n27;4;60;7;66175;29;65.74\n40;9;66;6;27805;16;70.45\n58;6;87;6;25237;19;94.73\n38;9;73;6;40056;24;72.49\n42;8;54;3;65543;21;45.11\n30;6;83;6;76556;12;103.29\n30;8;55;8;23343;18;59.43\n43;7;71;2;33500;35;61.55\n55;1;60;5;73222;11;75.24\n59;0;97;1;49375;18;102.39\n43;6;65;1;29662;11;59.55\n22;6;82;5;36964;8;79.11\n41;7;58;3;79638;6;55.45\n21;4;55;4;93666;27;42.49\n43;2;65;3;87215;13;63.35\n63;7;78;1;89042;30;89.7\n49;5;52;1;33284;18;57.87\n57;2;69;5;92789;15;101.19\n21;0;85;6;81389;4;104.3\n40;2;68;3;29435;34;80.91\n52;4;75;9;74340;11;92.68\n31;2;52;5;64078;24;70.0\n41;0;68;8;98832;20;112.9\n63;4;69;1;71293;35;77.23\n44;9;81;5;98781;22;87.21\n46;6;56;3;80403;15;64.66\n61;6;90;1;49124;38;107.22\n47;8;82;4;63919;38;78.32\n35;9;89;5;55247;13;97.24\n34;9;88;7;82752;30;86.65\n63;2;67;

In [6]:
!mkdir -p data/raw
!mv -f dataset_regresion_multiple_TA.csv data/raw/ 2>/dev/null || true

!ls -la data/raw



total 12
drwxr-xr-x 2 root root 4096 Sep 25 00:47 .
drwxr-xr-x 6 root root 4096 Sep 25 00:35 ..
-rw-r--r-- 1 root root 2558 Sep 25 00:37 dataset_regresion_multiple_TA.csv
-rw-r--r-- 1 root root    0 Sep 25 00:35 .gitkeep


In [7]:
%%writefile requirements.txt
pandas>=2.2
numpy>=1.26
scikit-learn>=1.4
statsmodels>=0.14
matplotlib>=3.8
seaborn>=0.13


Overwriting requirements.txt


In [8]:
!pip install -r requirements.txt


In [9]:
import os, csv
import pandas as pd

In [14]:
path = "data/raw/dataset_regresion_multiple_TA.csv"

In [15]:
# Autodetectar separador y cargar
with open(path, "r", encoding="utf-8") as f:
    sep = csv.Sniffer().sniff(f.read(4096), delimiters=[",",";","|","\t"]).delimiter

In [16]:
df = pd.read_csv(path, sep=sep)
df.columns = [c.strip().lower() for c in df.columns]

cols = ["edad","horas_ejercicio","peso","estres","ingresos","horas_tv","presion_arterial"]

print("Archivo:", path, "| separador:", repr(sep), "| shape:", df.shape)
df.head()

Archivo: data/raw/dataset_regresion_multiple_TA.csv | separador: ';' | shape: (100, 7)


edad  horas_ejercicio  peso  estres  ingresos  horas_tv  presion_arterial
0    58                7    81       1     21150        12             73.83
1    48                2    55       1     94740        22             82.26
2    34                2    81       3     86617        24             90.47
3    62                0    53       6     36896        34             93.49
4    27                4    60       7     66175        29             65.74

**1. Exploración inicial:**

-Describir el dataset (medias, máximos, mínimos, correlaciones).

-Identificar qué variables podrían estar más relacionadas con la presión arterial.

In [17]:
import pandas as pd

# columnas esperadas
cols = ["edad","horas_ejercicio","peso","estres","ingresos","horas_tv","presion_arterial"]
assert set(cols).issubset(df.columns), f"Faltan columnas en el dataframe: {set(cols)-set(df.columns)}"

# resumen estadístico básico
resumen = df[cols].describe().T[["mean","std","min","max"]]
resumen = resumen.rename(columns={"mean":"media","std":"desvio_std"}).round(3)

# nulos
nulos = df[cols].isna().sum().to_frame("nulos")

print("✓ Resumen (media, desvío, mínimo, máximo):")
display(resumen)

print("\n✓ Nulos por columna:")
display(nulos.T)


✓ Resumen (media, desvío, mínimo, máximo):


media  desvio_std       min       max
edad                 41.600      13.311     20.00     64.00
horas_ejercicio       4.370       3.014      0.00      9.00
peso                 73.280      14.299     50.00     99.00
estres                4.700       2.513      1.00      9.00
ingresos          59778.350   23603.253  20412.00  99605.00
horas_tv             19.760      11.137      0.00     39.00
presion_arterial     85.395      20.532     38.14    137.27


✓ Nulos por columna:


edad  horas_ejercicio  peso  estres  ingresos  horas_tv  \
nulos     0                0     0       0         0         0   

       presion_arterial  
nulos                 0

In [18]:
target = "presion_arterial"
corr = df[["edad","horas_ejercicio","peso","estres","ingresos","horas_tv",target]].corr()
corr_target = corr[target].drop(target).sort_values(ascending=False)
corr_abs = corr_target.abs().sort_values(ascending=False)
corr_target, corr_abs


(peso               0.588843
 edad               0.418731
 estres             0.366368
 horas_tv           0.040789
 ingresos          -0.009394
 horas_ejercicio   -0.358865
 Name: presion_arterial, dtype: float64,
 peso               0.588843
 edad               0.418731
 estres             0.366368
 horas_ejercicio    0.358865
 horas_tv           0.040789
 ingresos           0.009394
 Name: presion_arterial, dtype: float64)

**Descripción del dataset:**

**edad:** media 41.6 años (20–64). Rango razonable para adultos.

**horas_ejercicio:** media 4.37 h/sem (0–9). Hay personas que no hacen ejercicio (mín 0).

**peso:** media 73.28 kg (50–99). Variabilidad moderada.

**estrés:** media 4.7 (escala 1–10). Rango amplio (1–9).

**ingresos:** media 59778,35 (desvío 23603,25). Variabilidad alta.

**horas_tv:** media 19.76 h/sem (0–39). Variabilidad alta, también hay cero.

**presión arterial:** media 85.39 mmHg (38.14–137.27).

Nulos: 0 en todas las columnas, por lo que no hace falta imputación.

**Descripción con correlaciones:**

-La variable que más se asocia con la presión arterial es el **peso (r = 0.5888)**, relación moderada a fuerte y positiva: a mayor peso, mayor presión arterial.

-Le siguen la **edad (r = 0.4187)** y el **estrés (r = 0.3663)**, ambas también positivas y de magnitud moderada.

-Las **horas de ejercicio** muestran correlación negativa moderada **(r = –0.3588)**: a más ejercicio, menor presión arterial.

-En cambio, **horas de TV (r = 0.0407)** e **ingresos (r = -0.0093)** prácticamente no se relacionan con la presión arterial.

**2. Modelo de regresión lineal múltiple:**

-Estimar un modelo con todas las variables predictoras.

In [19]:
from sklearn.model_selection import train_test_split

predictoras = ["edad","horas_ejercicio","peso","estres","ingresos","horas_tv"]
objetivo = "presion_arterial"

X = df[predictoras].copy()
y = df[objetivo].copy()

X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.2, random_state=42)

X_tr.shape, X_te.shape


((80, 6), (20, 6))

In [20]:
from sklearn.linear_model import LinearRegression

modelo_full = LinearRegression()
modelo_full.fit(X_tr, y_tr)

print("Ajustado con todas las variables. Intercepto:", modelo_full.intercept_)


Ajustado con todas las variables. Intercepto: -19.895962407050817


In [21]:
import numpy as np
import pandas as pd

coef_tabla = pd.DataFrame({
    "variable": predictoras,
    "coeficiente": modelo_full.coef_
}).assign(abs_coef=lambda d: np.abs(d["coeficiente"])) \
  .sort_values("abs_coef", ascending=False) \
  .drop(columns="abs_coef")

coef_tabla


variable  coeficiente
3           estres     3.566095
1  horas_ejercicio    -1.606815
2             peso     0.796449
0             edad     0.704516
5         horas_tv     0.133066
4         ingresos     0.000098

Se construyó un modelo de regresión lineal múltiple utilizando las seis variables predictoras: edad, horas de ejercicio, peso, nivel de estrés, ingresos y horas de TV, con el objetivo de predecir la presión arterial.

El intercepto del modelo es aproximadamente **-19.8959**, lo cual representa el valor esperado de presión arterial cuando todas las variables predictoras son cero.

Los coeficientes indican la influencia de cada variable sobre la presión arterial, manteniendo las demás constantes:

**-Estrés (3.5660):** es la variable con mayor peso positivo, lo que sugiere que a mayor nivel de estrés, mayor presión arterial.

**-Horas de ejercicio (-1.6068):** tiene un efecto negativo importante, es decir, más ejercicio se asocia con menor presión arterial.

**-Peso (0.7964)** y **Edad (0.7045)**: presentan también un efecto positivo, aunque más moderado.

**-Horas de TV (0.1330):** tiene un impacto muy bajo en comparación con las anteriores.

**-Ingresos (0.000098):** prácticamente no tienen influencia sobre la presión arterial en este modelo.

En síntesis, el modelo sugiere que los factores más determinantes en la presión arterial son el estrés, el ejercicio, el peso y la edad, mientras que los ingresos y las horas de TV parecen tener un efecto mínimo.

**3. Selección de variables:**

-Determinar qué variables tienen mayor peso explicativo en la presión arterial.

-Comparar el modelo con todas las variables vs. uno simplificado con las más relevantes (analizar correlación por ejemplo u otro parámetro que consideres)

In [22]:
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
import pandas as pd
import numpy as np

# OLS para p-valores (en el set de entrenamiento)
X_tr_const = sm.add_constant(X_tr, has_constant='add')
ols_full = sm.OLS(y_tr, X_tr_const).fit()

pvals = ols_full.pvalues.drop("const", errors="ignore").sort_values()
pvals.name = "p_value"
display(pvals.to_frame())

# VIF sobre X_tr
X_vif = sm.add_constant(X_tr, has_constant='add')
vif_tab = pd.DataFrame({
    "variable": X_vif.columns[1:],
    "VIF": [variance_inflation_factor(X_vif.values, i) for i in range(1, X_vif.shape[1])]
}).sort_values("VIF", ascending=False)
display(vif_tab)


p_value
peso             3.981327e-14
edad             3.805822e-11
estres           1.089980e-10
horas_ejercicio  1.915131e-04
ingresos         5.466878e-02
horas_tv         2.347711e-01

variable       VIF
0             edad  1.090439
3           estres  1.069109
1  horas_ejercicio  1.057437
5         horas_tv  1.048260
4         ingresos  1.020230
2             peso  1.009608

In [24]:
# candidatas por p-valor
sig = pvals[pvals < 0.05].index.tolist()

# filtrar por VIF
keep = []
for v in sig:
    if v in vif_tab["variable"].values:
        vif_val = float(vif_tab.loc[vif_tab["variable"] == v, "VIF"].iloc[0])
    else:
        vif_val = 1.0
    if vif_val <= 10:
        keep.append(v)

selected = keep


In [25]:
print("Variables seleccionadas:", selected)


Variables seleccionadas: ['peso', 'edad', 'estres', 'horas_ejercicio']


In [26]:
from sklearn.linear_model import LinearRegression
import numpy as np
import pandas as pd

modelo_simple = LinearRegression().fit(X_tr[selected], y_tr)

coef_simple = pd.DataFrame({
    "variable": selected,
    "coeficiente": modelo_simple.coef_
}).assign(abs_coef=lambda d: np.abs(d.coeficiente)) \
 .sort_values("abs_coef", ascending=False) \
 .drop(columns="abs_coef")

print("Modelo simplificado ajustado con:", selected)
display(coef_simple)


Modelo simplificado ajustado con: ['peso', 'edad', 'estres', 'horas_ejercicio']


variable  coeficiente
2           estres     3.374270
3  horas_ejercicio    -1.597435
0             peso     0.788614
1             edad     0.699636

**Selección de variables:**

Criterio utilizado: Se estimó un OLS sobre el set de entrenamiento para obtener p-valores y se calculó VIF para chequear colinealidad. Regla de selección: p < 0.05 y VIF ≤ 10.

**Resultados (p-valores):**

**-Significativas:** peso (p=3.981327e-14), edad (p=3.805822e-11), estrés (p=1.089980e-10), horas_ejercicio (p=1.915131e-04).

**-No significativas:** ingresos (p=5.466878e-02), horas_tv (p=2.347711e-01).

**Colinealidad (VIF):** Todos los VIF 1.00–1.09, sin colinealidad significativa.

**Variables seleccionadas:** peso, edad, estres, horas_ejercicio

**Modelo simplificado (coeficientes):**

estrés = +3.374

horas_ejercicio = −1.597

peso = +0.788

edad = +0.699

**Interpretación:** Con control del resto de las variables, la presión arterial aumenta principalmente con estrés, peso y edad, y disminuye con más horas de ejercicio. Ingresos y horas de TV no mostraron evidencia estadística suficiente, por lo que se excluyeron del modelo.

**4. Evaluación del modelo:**

-Calcular el R², RMSE y MAE.

-Interpretar si el modelo ajusta adecuadamente.

In [28]:
import numpy as np
import pandas as pd
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

# Predicciones
yhat_full   = modelo_full.predict(X_te)
yhat_simple = modelo_simple.predict(X_te[selected])

# Baseline: predecir media del set de entrenamiento
yhat_base = np.full_like(y_te, fill_value=y_tr.mean(), dtype=float)

def calc_metrics(y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)   # sin 'squared'
    rmse = np.sqrt(mse)
    mae  = mean_absolute_error(y_true, y_pred)
    r2   = r2_score(y_true, y_pred)
    return {"R2": r2, "RMSE": rmse, "MAE": mae}

metrics = pd.DataFrame({
    "baseline_media":     calc_metrics(y_te, yhat_base),
    "modelo_completo":    calc_metrics(y_te, yhat_full),
    "modelo_simplificado":calc_metrics(y_te, yhat_simple),
}).T.round(4)

metrics


R2     RMSE      MAE
baseline_media      -0.0810  19.9277  16.7261
modelo_completo      0.7077  10.3624   8.5999
modelo_simplificado  0.7426   9.7249   8.2281

In [29]:
std_y = y_te.std()

def show(name):
    r2, rmse, mae = metrics.loc[name, ["R2","RMSE","MAE"]]
    return f"{name}: R²={r2:.3f}, RMSE={rmse:.2f}, MAE={mae:.2f}"

print("Desempeño (↑R² mejor, ↓RMSE/MAE mejor):")
print(show("baseline_media"))
print(show("modelo_completo"))
print(show("modelo_simplificado"))
print(f"Desvío estándar de y_test: {std_y:.2f} mmHg")


Desempeño (↑R² mejor, ↓RMSE/MAE mejor):
baseline_media: R²=-0.081, RMSE=19.93, MAE=16.73
modelo_completo: R²=0.708, RMSE=10.36, MAE=8.60
modelo_simplificado: R²=0.743, RMSE=9.72, MAE=8.23
Desvío estándar de y_test: 19.66 mmHg


La comparación de métricas muestra que tanto el modelo completo como el simplificado superan ampliamente al baseline, que apenas predice la media y no captura variabilidad (R² negativo).

El modelo completo explica aproximadamente un 71% de la variabilidad de la presión arterial (R²=0.7077) y reduce a la mitad el error respecto al baseline (RMSE = 10.3624 vs. 19.9277).

El modelo simplificado, con menos predictores, logra incluso un desempeño levemente superior (R²= 0.7426). Esto indica que las variables seleccionadas (peso, edad, estrés y horas_ejercicio) concentran la mayor capacidad explicativa del fenómeno.

En términos prácticos, el error medio (MAE = 8.2281 mmHg) es relativamente bajo comparado con el desvío estándar de la presión arterial en el set de prueba (19.66 mmHg), lo que sugiere un ajuste adecuado y clínicamente interpretable.

**Conclusión:** El modelo ajusta bien, siendo suficiente usar el modelo simplificado sin pérdida de desempeño, lo que además facilita su interpretación y aplicación.

**5. Informe final:**

El análisis de regresión lineal múltiple permitió identificar los principales factores que influyen en la presión arterial a partir de las variables estudiadas.

**Factores influyentes:**

**-Estrés:** Resultó ser la variable con mayor peso explicativo positivo. A mayor nivel de estrés, mayor es la presión arterial, lo que concuerda con la evidencia médica sobre el impacto del estrés en la salud cardiovascular.

**-Horas de ejercicio:** Mostró un efecto negativo significativo, es decir, más horas de actividad física se asocian con menores niveles de presión arterial.

**-Peso corporal:** También se relacionó de manera positiva con la presión arterial: a mayor peso, mayor presión.

**-Edad:** Tuvo un efecto positivo moderado, indicando que la presión arterial tiende a aumentar con la edad.

**Factores con menor influencia:**

**Ingresos** y **horas de televisión** no presentaron una importancia explicativa significativa en el modelo. Aunque podrían tener alguna relación indirecta (por ejemplo, a través de hábitos de vida asociados), en este dataset no se observa un aporte relevante para predecir la presión arterial.

**Comparación de modelos:**

El modelo completo (con todas las variables) explicó un 71% de la variabilidad de la presión arterial (R²=0.7077). Sin embargo, el modelo simplificado, que incluye sólo peso, edad, estrés y horas de ejercicio, alcanzó un desempeño incluso superior (R²=0.7426). Esto confirma que estas cuatro variables son suficientes para captar la mayor parte del fenómeno.

**Conclusiones:**

-Los factores peso, edad, nivel de estrés y ejercicio físico son los determinantes más importantes de la presión arterial en este análisis.

-Variables socioeconómicas como ingresos o hábitos pasivos como mirar televisión no tuvieron impacto relevante.

-La simplificación del modelo mejora su interpretabilidad y facilita la aplicación práctica en contextos de salud.

En síntesis, la presión arterial está fuertemente determinada por factores fisiológicos y de estilo de vida, siendo clave controlar el peso, reducir el estrés y mantener una rutina regular de ejercicio físico para prevenir niveles elevados.